src/Plugin/SmsGateway/Twilio.php

```diff
@@ -166,4 +174,61 @@ public function incoming(SmsMessageInterface $sms_message) {
       ->setReports([$report]);
   }
 
+  /**
+   * Validates the webhook request and creates an SMS message object.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *
+   * @return \Drupal\sms\Message\SmsMessage
+   */
+  protected function buildIncomingFromRequest(Request $request) {
+    $result = new SmsMessageResult();
+    $params = $request->request->all();
+    $report = (new SmsDeliveryReport())
+      ->setRecipient($params['To'])
+      ->setStatus(SmsMessageReportStatus::DELIVERED);
+    $sms = (new SmsMessage())
+      ->setMessage(trim($params['Body']))
+      ->setDirection(Direction::INCOMING)
+      ->setOption('data', $params)
+      ->setSenderNumber($params['From'])
+      ->addRecipients([$params['To']]);
+    if ($files = SmsUtility::processMedia($params)) {
+      $sms->setOption('media', $files);
+    }
+    try {
+      SmsUtility::validate($request, $this);
+    } catch (\Exception $e) {
+      $report->setStatus(SmsMessageReportStatus::REJECTED);
+      $report->setStatusMessage($e->getMessage());
+      $result->setError($e->getCode());
+      $result->setErrorMessage($e->getMessage());
+    }
+    $result->setReports([$report]);
```

Comment:
`setReports` can be cleaned up a lot by using `addReport` instead. See https://www.drupal.org/node/2836114